Shaking screen (fixes #2875)

raphaelsetin opened the pull request:
This solution fixes the bug when a user put their mouse cursor on the side of the map, the map begins to shake when it reaches its limit.
**CAUSE:** The problem was caused because the game first moves the map to where the mouse cursor is pointing, and then checks if the map is outside of the predefined boundaries. If the map is outside of the boundaries, then the game moves the map back inside the boundary, causing a shaking artifact for the user.
**OUR FIX:** Our solution first detects if the map-move will surpass its boundaries. If so, then the map is not moved. If the map-move doesn't surpass the boundaries, then the map is moved; and thus the final map position will for sure be inside the boundary.

Gymnasiast commented:
Could you rebase this against our develop branch?

raphaelsetin commented:
Sure @Gymnasiast I will soon!

raphaelsetin commented:
@Gymnasiast @IntelOrca I just rebased!

Goddesen commented:
Fixes shaking for cursor edge scrolling, as far as I could test it. Some shaking is still evident when gliding along the edge of the map, though that was already a thing.

Should ideally also handle hotkey and mouse scrolling.

raphaelsetin commented:
@Goddesen Could you maybe upload a video of it or explain a little in depth for me? Because I didn't understand when you said "Some shaking is still evident when gliding along the edge of the map".
Maybe you meant when the map is being moved (still inside the boundary), it kind of moves in an "unsmooth" fashion? If that's what you meant, then I agree; still, that wouldn't be in the scope of this pull request btw, even though it's nice to point that out.

And you're right about the hotkey scrolling and mouse scrolling, I haven't tested for those. In principle, this solution was only meant for mouse cursor edge scrolling, since that's the issue I found here in the repo. But thanks nonetheless for pointing that out.

Goddesen commented:
@raphaelsetin In [this video](https://www.youtube.com/watch?v=CYcEiNJmaGU), at 13s I am scrolling towards the right, causing the camera view to glide along the edge of the map. The screen is visibly shaking.

raphaelsetin commented:
@Goddesen You were definitely right in relation to the keyboard keys bug. I just corrected and rebased my branch.

In regarding of the video, I definitely see what you mean there, but I don't think the scope of this pull request goes to this as well. Still, that's something I can maybe address in either this pull request, or another one, since it's more of a drawing-engine related issue I think.

The shaking while gliding it's in my opinion, since I didn't investigate in depth, in relation to the way OpenRCT2 draws the viewport AND the spaces it uses to draw (i.e. if it wants to move in diagonal, it draws 12 pixels up and 12 pixels right, making it seems it's not smooth). I played with this setting (magic number 12 in the code btw), and if I lower that, the movement of the map gets too slow, so the fix is maybe in another place. Again, I'm not sure if it's related to the same issue that I'm addressing in this pull request, but thank you nonetheless for your input, it helped for sure.

@IntelOrca @Gymnasiast I just rebased too, let me know if I'm still missing anything.

raphaelsetin commented:
@Gymnasiast @IntelOrca Does this new solution seems good for you? If not, let me know that I can change it.

IntelOrca commented:
It still shakes for me.
If you go into Crazy Castle then hold down the right arrow key, it will move up the edge the map shaking the whole way rather than smoothly.

Gymnasiast commented:
Superseded by #6620.
